File: scripts/split-all-json.mjs
/**
 * Split src/data/generated/all.json into <45MB chunk files so the catalogue can
 * live in the GitHub repo (GitHub rejects any single file > 100MB and warns
 * above 50MB).
 *
 * The chunks are RAW BYTE slices (not JSON-aware) named all.json.part-000,
 * all.json.part-001, ... plus a manifest all.json.parts.json listing them in
 * order. `join-all-json.mjs` concatenates them back into an identical all.json
 * at the start of every build.
 *
 * STREAMING: the 140MB+ file cannot be read into a single Buffer on the
 * low-memory sandbox (OOM). We read it as a stream and cut a new part every
 * CHUNK bytes, so peak memory stays tiny.
 *
 * Run manually whenever all.json changes:  node scripts/split-all-json.mjs
 */
import {
  createReadStream,
  createWriteStream,
  writeFileSync,
  readdirSync,
  unlinkSync,
  statSync,
} from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = join(dirname(fileURLToPath(import.meta.url)), '..');
const genDir = join(root, 'src/data/generated');
const ALL = join(genDir, 'all.json');
const CHUNK = 45 * 1024 * 1024; // 45MB per part — under GitHub's 50MB warning threshold

// clean old parts
for (const f of readdirSync(genDir)) {
  if (f.startsWith('all.json.part-')) unlinkSync(join(genDir, f));
}

const totalBytes = statSync(ALL).size;

async function main() {
  const parts = [];
  let idx = 0;
  let written = 0; // bytes written into the current part
  let ws = null;

  const openPart = () => {
    const name = `all.json.part-${String(idx).padStart(3, '0')}`;
    parts.push(name);
    ws = createWriteStream(join(genDir, name));
    written = 0;
  };
  const closePart = () =>
    ws ? new Promise((res, rej) => ws.end((e) => (e ? rej(e) : res()))) : Promise.resolve();

  const writeChunk = (b) =>
    new Promise((res, rej) => {
      if (ws.write(b)) res();
      else {
        ws.once('drain', res);
        ws.once('error', rej);
      }
    });

  openPart();

  const rs = createReadStream(ALL, { highWaterMark: 8 * 1024 * 1024 });
  for await (const chunk of rs) {
    let offset = 0;
    while (offset < chunk.length) {
      const room = CHUNK - written;
      const take = Math.min(room, chunk.length - offset);
      const slice = chunk.subarray(offset, offset + take);
      await writeChunk(slice);
      written += take;
      offset += take;
      if (written >= CHUNK) {
        await closePart();
        idx++;
        openPart();
      }
    }
  }
  await closePart();

  writeFileSync(
    join(genDir, 'all.json.parts.json'),
    JSON.stringify({ parts, bytes: totalBytes }, null, 2)
  );
  console.log(`[split] all.json (${totalBytes} bytes) -> ${parts.length} parts:`, parts.join(', '));
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: scripts/split_all_json.py
"""
Low-RAM byte-level splitter for all.json -> all.json.part-NNN (45MB each) +
all.json.parts.json manifest. Mirrors scripts/split-all-json.mjs output exactly
(raw byte slices) but runs in Python with a tiny fixed buffer, which is far more
reliable on this 985MB-RAM sandbox than the Node streaming version.
"""
import os, json, glob

ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
GEN = os.path.join(ROOT, 'src/data/generated')
ALL = os.path.join(GEN, 'all.json')
CHUNK = 45 * 1024 * 1024
BUF = 4 * 1024 * 1024

# clean old parts
for f in glob.glob(os.path.join(GEN, 'all.json.part-*')):
    os.remove(f)

total = os.path.getsize(ALL)
parts = []
idx = 0
written = 0
out = None


def open_part():
    global out, idx, written
    name = f'all.json.part-{idx:03d}'
    parts.append(name)
    out = open(os.path.join(GEN, name), 'wb')
    written = 0


open_part()
with open(ALL, 'rb') as f:
    while True:
        need = min(BUF, CHUNK - written)
        data = f.read(need)
        if not data:
            break
        out.write(data)
        written += len(data)
        if written >= CHUNK:
            out.close()
            idx += 1
            open_part()
out.close()

# drop a trailing empty part if the file ended exactly on a boundary
last = os.path.join(GEN, parts[-1])
if os.path.getsize(last) == 0 and len(parts) > 1:
    os.remove(last)
    parts.pop()

json.dump({'parts': parts, 'bytes': total},
          open(os.path.join(GEN, 'all.json.parts.json'), 'w'), indent=2)
print(f'[split] all.json ({total} bytes) -> {len(parts)} parts: ' + ', '.join(parts))
